[V1][Metrics] Add API for accessing in-memory Prometheus metrics

markmc opened the pull request:
The V0 LLM offline inference API exposes per-request metrics via `RequestOutput.RequestMetrics`. In V1, so far we have chosen to not track per-request metrics or implement this API.

All recent work implementing EAGLE have been using `examples/offline_inference/eagle.py` which depends on these metrics to report an aggregated mean acceptance length number.

See e.g.  the EAGLE3 PR #16937 which used a WIP implementation of the per-request metrics #16367

The proposal in this PR is to achieve the same aggregated view by using the Prometheus metrics already implemented for the online serving case. This will mean we automatically gain new spec decoding metrics in #16665 for both offline and online inferencing

This does not preclude us from implementing per-request metrics in future in V1 if that proves to be important.

See also [the spec decoding metrics design doc](https://docs.google.com/document/d/1xoxP9X9Zt6wkpIhErUcI7g0XtiFtKkvu1Kjaxi_uJqU).

markmc committed:
[V1][Metrics] Remove _total suffix from Counters

prometheus_client does this automatically:

```
def _build_full_name(metric_type, name, namespace, subsystem, unit):
    ...
    if metric_type == 'counter' and full_name.endswith('_total'):
        full_name = full_name[:-6]  # Munge to OpenMetrics.
```

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Add API for accessing in-memory Prometheus metrics

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

luyuzhe111 commented:
@markmc Can you please provide an example of how to compute acceptance length from the retrieved metrics in `examples/offline_inference/eagle.py`? is it just

```
acceptance_length = 1 + (
    metrics.get_value('vllm:spec_decode_num_accepted_tokens') /
    metrics.get_value('vllm:spec_decode_num_drafts')
)
```
additionally, I'm wondering why `vllm:spec_decode_num_accepted_tokens_per_pos` is a counter instead of a vector? 

thanks again for the PR!

markmc commented:
> @markmc Can you please provide an example of how to compute acceptance length from the retrieved metrics in `examples/offline_inference/eagle.py`? is it just
> 
> ```
> acceptance_length = 1 + (
>     metrics.get_value('vllm:spec_decode_num_accepted_tokens') /
>     metrics.get_value('vllm:spec_decode_num_drafts')
> )
> ```
> 
> additionally, I'm wondering why `vllm:spec_decode_num_accepted_tokens_per_pos` is a counter instead of a vector? how is it defined?
> 
> thanks again for the PR!

All good questions, @luyuzhe111. In fact, I had already rebased onto main, tried to use `num_accepted_tokes_per_pos` via the API and saw this deficiency!

Try the new version. I've adopted your suggestion of adding a `Vector` abstraction as well :+1:

luyuzhe111 commented:
Hi @markmc, appreciate the fast turn-around! the new version works like a charm. the only request is to add plus 1 to the mean acceptance length since one token will always be accepted. so mean acceptance length is essentially "average number of tokens generated per forward pass".  cc @LiuXiaoxuanPKU

markmc commented:
>  the only request is to add plus 1 to the mean acceptance length since one token will always be accepted. so mean acceptance length is essentially "average number of tokens generated per forward pass". 

I don't think of the bonus/recovered token as "accepted", particularly in the context of the acceptance rate calculation - the proportion of drafts (speculated tokens) that are accepted

let's take [the example from here](https://github.com/vllm-project/vllm/pull/16665#issuecomment-2826563829)

```
num_spec_tokens = 3

drafts:
- #1: 3 accepted
- #2: 1 accepted
- #3: 2 accepted
- #4: 2 accepted
- #5: 1 accepted

observe:
- num_drafts = 5
- num_draft_tokens = 15
- num_accepted_tokens = 9
- accepted_tokens_per_pos = [5, 3, 1]

compute:
- acceptance_rate = 9/15 = 0.6
- mean_acceptance_length = 1.8
- acceptance_probs_per_pos = [1.0, 0.6, 0.2]
```

You want:

```
compute:
- acceptance_rate = 9/15 = 0.6
- mean_acceptance_length = 2.8
- acceptance_probs_per_pos = [1.0, 1.0, 0.6, 0.2]
```

Why? Got any references to show this being common practice? Thanks.

luyuzhe111 commented:
Hi @markmc, as far as I know, all speculative decoding literatures reporting acceptance length includes the bonus token since this quantity aligns with "number of tokens generated per forward pass". All EAGLE papers ([EAGLE-1](https://arxiv.org/abs/2401.15077), [EAGLE-2](https://arxiv.org/abs/2406.16858), [EAGLE-3](https://arxiv.org/abs/2503.01840))'s reported mean acceptance lengths include the +1 bonus token.

alternatively, maybe it's worth keeping the acceptance rate metrics as is but adding another metric for "number of tokens generated per forward pass"?

markmc commented:
> Hi @markmc, as far as I know, all speculative decoding literatures reporting acceptance length includes the bonus token since this quantity aligns with "number of tokens generated per forward pass". 

Ok, see #17908. Thanks!

LiuXiaoxuanPKU commented on vllm/v1/metrics/reader.py:
```diff
@@ -0,0 +1,160 @@
+                # We convert these into a vector of integer values.
+                #
+                assert metric.name == \
+                    "vllm:spec_decode_num_accepted_tokens_per_pos"
```
Nit: duplicate check, we checked in the if statement.

markmc replied:
Fixed, thanks!

LiuXiaoxuanPKU commented on examples/offline_inference/eagle.py:
```diff
@@ -92,30 +93,37 @@ def main():
+    try:
+        metrics = llm.get_metrics()
+    except AssertionError:
+        # Presumably V0 engine, metrics not supported
```
Can we just add a warning message here:
```suggestion
        print("Metrics are not supported in the V0 engine.")
```

LiuXiaoxuanPKU commented on tests/v1/engine/test_llm_engine.py:
```diff
@@ -97,3 +98,52 @@ def test_parallel_sampling(vllm_model, example_prompts) -> None:
+        assert "+Inf" in request_generation_tokens[0].buckets
+        assert request_generation_tokens[0].buckets["+Inf"] == n_prompts
+        assert request_generation_tokens[0].count == n_prompts
+        assert request_generation_tokens[0].sum == total_tokens
```
Add tests for `Vector` as well?

markmc replied:
Done now, in two places :+1:

LiuXiaoxuanPKU commented on vllm/v1/metrics/reader.py:
```diff
@@ -0,0 +1,160 @@
+            for s in samples:
+                collected.append(
+                    Counter(name=metric.name,
+                            labels=samples[0].labels,
```
QQ: Can we safely get the first sample here?

markmc replied:
Great question - basically a bug that you'd only see with a DP config (multiple samples labelled with different `engine_id`)

LiuXiaoxuanPKU reviewed (commented):
Thanks for the PR! Overall LGTM!
One concern/question I have is whether enabling log_stats for llm_engine could noticeably degrade performance from your experience?

markmc committed:
[V1][Metrics] Address API PR feedback

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Add test for Vector

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Account for multi-engine DP configuration

In the case of DP, we will have a complete set of metrics for
each DP rank.

We could make get_metrics_snapshot() take a DP rank parameter
to avoid this, but it is possible in future we will add further
dimensions that we want to label on.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Add unit test for metrics reading API

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented on vllm/v1/metrics/loggers.py:
```diff
@@ -332,12 +332,12 @@ def __init__(self, vllm_config: VllmConfig, engine_index: int = 0):
         #
         self.counter_spec_decode_num_draft_tokens = \
```
Fixed now I think

WoosukKwon commented:
@markmc Is this PR waiting for review? Or is it in progress?

markmc commented:
> @markmc Is this PR waiting for review? Or is it in progress?

It is waiting for review

LiuXiaoxuanPKU commented:
LGTM, @markmc  could you just double check if the CI failure is related so that we can merge this PR?

markmc commented:
> LGTM, @markmc could you just double check if the CI failure is related so that we can merge this PR?

Yes, AFAICT all of these failures are happening on other PRs too

WoosukKwon commented:
@markmc Can you please merge from main again?

markmc commented:
> @markmc Can you please merge from main again?

Done. I don't think the rebase resolves any of the test failures, but I could be wrong

markmc committed:
[V1][Metrics] Fix docs build

```
docs/source/serving/engine_args.md:14: ERROR: Failed to import "_engine_args_parser" from "vllm.engine.arg_utils".
No module named 'prometheus_client'
```

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
Ok, the docs failure was a genuine - but hard-to-spot - issue with the PR

```
vllm/docs/source/serving/engine_args.md:14: ERROR: Failed to import "_engine_args_parser" from "vllm.engine.arg_utils".
No module named 'prometheus_client'
```
